### DISCUSSION TITLE
How to calculate tax rate from total & tax amounts using allocate?

### DISCUSSION THREAD

**Discussion by CanRau** [dinerojs/dinero.js#712] [Q&A] [answered]
I do currently `amount_tax / (amount_total - amount_tax)` or `amount_tax / price_net`

and I'm trying to convert to use dinero, which works great for the most part only divisions I don't get yet or rather how to do them in this case?

**Comment by AbdlrahmanSaberAbdo** [comment]
> I do currently `amount_tax / (amount_total - amount_tax)` or `amount_tax / price_net`
> 
> and I'm trying to convert to use dinero, which works great for the most part only divisions I don't get yet or rather how to do them in this case?

Check this, could help https://v2.dinerojs.com/docs/faq/how-do-i-calculate-a-percentage

**Reply by whitespacecode** [reply]
I'm having the same issue, using allocate or multiply is resulting in different values than what you suppose to have when trying to resolve the tax amounts.

Example:
```
const priceAmount = 9700; 
const tax = 21;

const price = dinero({
   amount: priceAmount.value,
   currency: EUR,
});

const multiplied = multiply(priceAmount, { amount: tax, scale: 2 });
const [d1, d2] = allocate(priceAmount, [value, (100 - tax)]); 

toSnapshot(multiplied); //20,37
toSnapshot(d1, d2); //20,37  76,63
```

Expected result:
```
const subtotal = priceAmount / 1 + (tax / 100);
const taxAmount = (priceAmount * tax / 100) / tax;

console.log(subtotal); //80,17
console.log(taxAmount); //16,83
```

**Comment by CanRau** [comment]
It's been a while, sorry, I'm currently struggling with a slightly different calculation, rather than calculating the tax rate I'm trying to get the included tax amount (as typical in germany)
You'd do `amount / 1.19 * 0,19` though even changing the scale and calculate with 119 & 19 I can't get it to divide by those amounts or don't understand how to do it properly

**Reply by AbdlrahmanSaberAbdo** [reply]
How do you save taxes in your DB?

Based on my experience, I always use the 'dinero' for calculations. what I typically do is the following:

- Calculate the scaled tax amount, add 2 to the specified scaling unit (since taxes are saved as a percentage) and sum it by the scaling unit value. For example, if the tax value is `1.553`, and your scaling unit is `5` (scaling unit is determined by you, it depends on how decimal numbers you can get for tax e.g. 1.555, 1.1234567, etc..), so the value will be saved in the database as `155300` (here I'm assuming that you saved the tax value in db by fixed scale (scale unit for tax) ). When retrieving the value from the database, the scaled amount will be as follows:

```
{
  amount: taxValue,
  scale: 5 (fixed scale value for tax) + 2
}
```
so the value will be `0.01553`

- Determine the tax amount by multiplying the total price with the scaled tax amount using the `di.multiply` function.
- Add the calculated tax amount to the total price using the `di.add` function to get the final price including taxes.

**Comment by sarahdayan** [comment, marked as answer]
Hey @CanRau! A tax rate is a ratio, not a monetary value, so you wouldn't use Dinero objects for the result. You'd extract the amounts, do the division with plain numbers, and keep using Dinero for the money parts.

For extracting included tax (your German VAT example), you can use [`multiply`](https://v2.dinerojs.com/api/mutations/multiply.html):

```js
import { dinero, multiply } from 'dinero.js';
import { EUR } from 'dinero.js/currencies';

const gross = dinero({ amount: 11900, currency: EUR }); // €119.00 (incl. 19% VAT)
const tax = multiply(gross, { amount: 19, scale: 119 }); // 19/119 of gross
```

The [calculating percentages](https://v2.dinerojs.com/guides/calculating-percentages.html) guide covers this pattern in more detail.
